### PR TITLE
Add support for account creation through SSO provider

### PR DESCRIPTION
**Pull request by dwesty17**
Basically exactly what @benhaynes mentioned [here](https://github.com/directus/v8-archive/issues/664#issuecomment-445570679).

If a user tries to authenticate through an SSO provider for the first time, directus will create a new account for them if both of the following are true...
1. There isn't already a record in the `dierctus_users` table with matching email
2. Their email address is on a domain whitelist that is configurable through the `OAUTH_USER_CREATION_DOMAIN_WHITELIST ` environment variable

I haven't added any tests or docs yet, but I'm happy to if ya'll think this approach works and are open to accepting the PR. This would be a super helpful feature for a relatively large company like ours that wants to use Directus 🙏

### CODE REVIEW

**Comment by dwesty17** [comment]
This is basically and attempt the MVP part of https://github.com/directus/directus/discussions/3956

**Comment by benhaynes** [comment]
This is awesome, thank you @dwesty17!! Can't wait to check this one out... really useful feature. Does this support multiple allow-listed domains? eg: CSV

**Comment by dwesty17** [comment]
Happy to help! Had a good time with it. 

Also - yep. The code expects a comma separated list of acceptable domains.

**Comment by rijkvanzanten** [comment]
I'm wondering if we should make this one "scope" bigger, and consider it an allowlist for registration an sich, instead of just registration through oAuth. 

Being able to publicly register is being discussed in https://github.com/directus/directus/discussions/3083; "domain allowlist", could be a setting in there, and enabling registration on the "global level" would effectively enable registration through the API _and_ oAuth at the same time 🤔

**Comment by tylerforesthauser** [comment]
> I'm wondering if we should make this one "scope" bigger, and consider it an allowlist for registration an sich, instead of just registration through oAuth.
> 
> Being able to publicly register is being discussed in #3083; "domain allowlist", could be a setting in there, and enabling registration on the "global level" would effectively enable registration through the API _and_ oAuth at the same time 🤔

Has there been any movement on this? I would love to see this feature built-in, rather than trying to spin this up myself.

**Comment by rijkvanzanten** [comment]
@tylerforesthauser No, there's a couple unanswered questions left to resolve in that discussion.. I am pretty convinced that we should have a single user-registration configuration that works for both "regular" and oAuth-based signups though.

As of right now, I'm thinking we add "allow user registration", "user registration email allow list", and "user registration role" as options to project settings (directus_settings), and have that enable both a "regular" register flow and the ability to automatically create an account through oAuth if you login with that (cc @benhaynes)

**Comment by benhaynes** [comment]
Absolutely. We need to keep this code consolidated, and I think Registration should be global. Having this work for "normal" registering and SSO seems like the ideal path forward. Those three Project Setting options look good to me! 👍

File: api/src/controllers/auth.ts [hunk 1]
```diff
@@ -253,11 +253,20 @@ router.get(
...
 			schema: req.schema,
 		});
+		const userService = new UsersService({ schema: req.schema });
 
 		const email = getEmailFromProfile(req.params.provider, req.session.grant.response?.profile);
 
 		req.session?.destroy(() => {});
 
+		const whitelistedDomains = process.env.OAUTH_USER_CREATION_DOMAIN_WHITELIST?.split(",").map(d => d.trim());
+		const emailDomainIsWhitelisted = whitelistedDomains?.some(d => email.endsWith(d)) ?? false;
```
Comment (xesjkeee) [inline api/src/controllers/auth.ts:263]: I think you could validate domains in whitelist for starting with `@`, otherwise if my whitelist would be like `bar.com`, I can simply bypass this condition with creating own `foobar.com` domain

**Comment by megamorf** [comment]
+1 for this feature. As it is right now newly created OIDC users will be greeted with an obscure error message after signing in with a configured SSO provider. It would be much appreciated if user accounts do not need to exist and be activated for SSO to work.

```
json
{"errors":[{"message":"Invalid user credentials.","extensions":{"code":"INVALID_CREDENTIALS"}}]}
```

**Comment by workatease** [comment]
is there a reason why the PR is not merged yet?

**Comment by rijkvanzanten** [comment]
@workatease see my previous comment: https://github.com/directus/directus/pull/3955#issuecomment-785426813

**Comment by workatease** [comment]
Sorry was just curious to see this feature

**Comment by rijkvanzanten** [comment]
Haha no problem! I'm not annoyed or anything, just pointing out that nothing changed since then 😉

File: api/src/controllers/auth.ts (hunk 1 above)
Comment (dwesty17) [reply api/src/controllers/auth.ts:263]: Nice catch! Thanks 🙏

**Comment by dwesty17** [comment]
Hey everyone sorry for the slow response. Got busy at work and then was on vacation skiing for a little bit. I do want to pick this back up, but it probably won't be till this weekend or next week.

I don't wanna be a blocker though, so if anyone wants to pick up where I left off by all means go for it.

**Comment by nichols-green** [comment]
I'd like to add my perspective to this. This and other projects that I've looked at often start with the assumption that users will be created for this system, and neglect the enterprise model of users existing outside, and the system simply connecting to that universe. 

I can also imagine that this type of feature would need to be quite flexible. We will need a way to add a custom script to parse information on registration. Not only do we need to parse the user when they first login, we may want to bypass the login page entirely and only allow users to login through the IDP. 

Finally, we have roles. In order to ensure the data remains protected, we need to assign roles dynamically on login. Each IDP and org will be different with different shapes and different data. So again, I see the need for an on-login script, and an on-registration script.

**Comment by alexisig** [comment]
Any idea if this PR is likely to be merged sometime soon?

I'm happy to contribute if it needs more work

**Comment by rijkvanzanten** [comment]
@alexisig Nothing has changed since my previous comment on the matter 🙂 :

https://github.com/directus/directus/pull/3955#issuecomment-785426813

TL;DR: I'd like to make this a "proper" User Registration feature, instead of just supporting automatic user registration through oAuth. 

In the meantime, you can use the `oauth` hook to automatically create and login a user if it doesn't exist yet

**Comment by alexisig** [comment]
@rijkvanzanten 

Thanks for the feedback  :slightly_smiling_face: 

Can you tell me what you think about the following breakdown of the feature?

Based on the three points you suggested:
1) Allow user registration (anyone can signup):
- oauth:
   - Automatic signup with the `Login with` button
   - User info is synced every time they log in (`OAUTH_SYNC_INFO_ON_LOGIN`, `true` by default)

- regular auth:
   - New view: signup form
   - New button: signup button on `admin/login` page

2) User registration email allow list (anyone with proper email domain name can signup):
- oauth only:
   - Automatic signup with the `Login with` button
   - Retrieved email domains must match one of the configured domain (with `OAUTH_USER_CREATION_DOMAIN_WHITELIST` ?)

3) User registration role
- Automatic import of the user's roles / groups with the info on login
- Admin interface to map external role name to directus role 
- Option to automatically create and map directus roles from external role name `OAUTH_AUTOMATIC_ROLE_MAPPING` (`false` by default)
- Roles claim name can be configured with `OAUTH_ROLE_CLAIM_NAME`

**Comment by rijkvanzanten** [comment]
Points 1 and 2 are exactly what I'm thinking. Could you elaborate a little on the points under 3?

Not exactly sure what you mean by the following:

> Automatic import of the user's roles / groups with the info on login
> Admin interface to map external role name to directus role
> Roles claim name can be configured with OAUTH_ROLE_CLAIM_NAME

**Comment by alexisig** [comment]
@rijkvanzanten 
Thanks for the quick feedback

1. Automatic import of the user's roles / groups with the info on login
-> So that directus users roles memberships are updated on each login, based on the user roles membership from the provider

2.  Admin interface to map external role name to directus role
-> I'm thinking about something similar to Nextcloud's social login's openid configuration panel:
image
Basically roles name from the provider can be matched to roles with different name on directus.

3.  Roles claim name can be configured with OAUTH_ROLE_CLAIM_NAME
-> Not sure this is relevant for all providers, I know it is for openid providers, but roles could come from a different claim than `role`, `groups` for example.

**Comment by rijkvanzanten** [comment]
Ah gotcha, that all makes sense 👍🏻 

As for `OAUTH_ROLE_CLAIM_NAME`, I think we should just rely on the regular scope request function that already exists, combined with a `OAUTH_ROLE_PATH` which is a dot-notation path of where in the provider's API response we'd find the role, so you can set it to `OAUTH_ROLE_PATH="profile.role.uid"` f.e.

**Comment by alexisig** [comment]
Sunds good, thanks! :)

I will get started on a different PR and let you know if more questions
come up

**Comment by rijkvanzanten** [comment]
Awesome! Thank you 🙂 ❤️

**Comment by nichols-green** [comment]
IMO It's very likely that users will want the flexibility to map/parse roles for each external provider. At the very lease keep it in a hook so I can switch on provider to write my mappings or other logic. If that becomes prohibitive to implementing this feature, I would think a single map would allow most people to get rolling with their OAuth integration, with the expectation that scoped mappings would be supported in the future.

**Comment by peteruithoven** [comment]
Just to save other people work, based on work in https://github.com/directus/directus/discussions/5379 and this pull request I created this hook to automatically create accounts:
``` js 
const whitelistedDomains = ["..."];

module.exports = function registerHook({ services, database, getSchema }) {
  const { UsersService } = services;
  return {
    "oauth.*.login.before": async function (oauthProfile) {
      const { email } = oauthProfile.profile;
      const schema = await getSchema();
      const usersService = new UsersService({
        schema,
        knex: database
      });

      const emailDomainIsWhitelisted = whitelistedDomains.some((d) =>
        email.includes(d)
      );
      if (!emailDomainIsWhitelisted) return;

      const existingUser = await database("directus_users")
        .where({ email })
        .first();
      if (existingUser) return;

      await usersService.createOne({
        email: email,
        role: "1f7e3cdd-b61c-4fb7-8c5a-4c509d52d41d"
      });
    }
  };
};
```

**Comment by JohannesDienst** [comment]
@alexisig Did you start a developing this feature?

We need this to synchronize/create Accounts with the roles we get in the OpenId-Token, so we are eager to help if needed :)

**Comment by rijkvanzanten** [comment]
This was implemented in https://github.com/directus/directus/pull/8660 in a bigger overhaul of the authentication flow, which also added support for LDAP. You can set [the `AUTH_<PROVIDER>_ALLOW_PUBLIC_REGISTRATION` environment variable](https://docs.directus.io/configuration/config-options/#oauth-2-0) to enable signup through OAuth2/OpenID 🎉
